src/platform/windows/misc.cpp

```diff
@@ -395,46 +468,113 @@ namespace platf {
-    HANDLE shell_token = duplicate_shell_token();
-    if (!shell_token) {
-      // This can happen if the shell has crashed. Fail the launch rather than risking launching with
-      // Sunshine's permissions unmodified.
-      ec = std::make_error_code(std::errc::no_such_process);
+  create_boost_child_from_results(bool process_launched, const std::string &cmd, std::error_code &ec, PROCESS_INFORMATION &process_info, bp::group *group) {
+    // Use RAII to ensure the process is closed when we're done with it, even if there was an error.
+    auto close_process_handles = util::fail_guard([process_launched, process_info]() {
+      if (process_launched) {
+        CloseHandle(process_info.hThread);
+        CloseHandle(process_info.hProcess);
+      }
+    });
+
+    if (ec) {
+      // If there was an error, return an empty bp::child object
       return bp::child();
     }
 
-    auto token_close = util::fail_guard([shell_token]() {
-      CloseHandle(shell_token);
-    });
+    if (process_launched) {
+      // If the launch was successful, create a new bp::child object representing the new process
+      auto child = bp::child((bp::pid_t) process_info.dwProcessId);
+      if (group) {
+        // If a group was provided, add the new process to the group
+        group->add(child);
+      }
 
-    // Populate env with user-specific environment variables
-    if (!merge_user_environment_block(env, shell_token)) {
-      ec = std::make_error_code(std::errc::not_enough_memory);
+      BOOST_LOG(info) << cmd << " running with PID "sv << child.id();
+      return child;
+    }
+    else {
+      auto winerror = GetLastError();
+      BOOST_LOG(error) << "Failed to launch process: "sv << winerror;
+      ec = std::make_error_code(std::errc::invalid_argument);
+      // We must NOT attach the failed process here, since this case can potentially be induced by ACL
+      // manipulation (denying yourself execute permission) to cause an escalation of privilege.
+      // So to protect ourselves against that, we'll return an empty child process instead.
       return bp::child();
     }
+  }
 
-    // Most Win32 APIs can't consume UTF-8 strings directly, so we must convert them into UTF-16
-    std::wstring wcmd = utf8_to_wide_string(cmd);
-    std::wstring env_block = create_environment_block(env);
-    std::wstring start_dir = utf8_to_wide_string(working_dir.string());
+  /**
+ * @brief Impersonate the current user, invoke the callback function, then returns back to system context.
+ *
+ * @param user_token A handle to the user's token that was obtained from the shell
+ * @param callback A function that will be executed while impersonating the user
+ * @return An std::error_code object that will store any error that occurred during the impersonation
+ */
+  std::error_code
+  impersonate_current_user(HANDLE user_token, std::function<void()> callback) {
```

Comment:
Since there's only one codepath where you need to impersonate now, you can put this back in `run_command()`.